### PR TITLE
Adds 'lib' reference directives

### LINKED ISSUE #15732: Suggestion: Add "lib" reference directive
Following discussion at https://github.com/DefinitelyTyped/DefinitelyTyped/issues/15140, it seems that we might benefit from adding a new `lib` reference directive similar to the `types` reference directive we have today. The `lib` reference directive would allow an individual file to specify one or more **lib** files to include when the file is included.

### Example
```ts
/// <reference lib="es2015, dom" />
```

This would allow individuals using `@types/core-js` (or any other polyfill/shim library) to automatically get the correct types.

### PR DESCRIPTION
**Pull request by rbuckton**
This change adds support for a `/// <reference lib="name" />` directive, allowing a file to explicitly include an existing built-in _lib_ file.

* When a built-in _lib_ is referenced via the `lib` reference directive, any `no-default-lib` reference directive in the lib file is ignored. This is so that a `lib` reference in an imported package won't cause your current project to lose its default lib if you haven't specified one.
* Built-in _lib_ files are referenced in the same fashion as the `"lib"` compiler option in _tsconfig.json_ (e.g. use `lib="es2015"` and not `lib="lib.es2015.d.ts"`, etc.).

In the long term this can help polyfill/shim packages like core-js or es6-shim.

NOTE: Apparently we were very close to running out of heap memory when running `gulp tests` after running `gulp clean` as gulp is holding onto too much memory. Individually each task is fine, but in concert they add up to a large amount of uncollected memory. To address this, I've added a script to run some of our builds out-of-process. This is not an issue in the `jake` builds.

Fixes #15732.
Supersedes #15780

### CODE REVIEW

File: src/compiler/parser.ts [hunk 1]
```diff
@@ -7613,13 +7615,17 @@ namespace ts {
...
                     const typeReferenceDirectives = context.typeReferenceDirectives;
+                    const libReferenceDirectives = context.libReferenceDirectives;
                     forEach(toArray(entryOrList), (arg: PragmaPsuedoMap["reference"]) => {
                         if (arg.arguments["no-default-lib"]) {
                             context.hasNoDefaultLib = true;
                         }
                         else if (arg.arguments.types) {
                             typeReferenceDirectives.push({ pos: arg.arguments.types.pos, end: arg.arguments.types.end, fileName: arg.arguments.types.value });
                         }
+                        else if (arg.arguments.lib) {
```
Comment (andy-ms) [inline src/compiler/parser.ts]: #23904 Could avoid repeated code here

Comment (rbuckton) [reply src/compiler/parser.ts]: Do you want me to do this or do you want to do in your other PR once this is merged?

Comment (andy-ms) [reply src/compiler/parser.ts]: I'll close the other PR since it doesn't have much benefit with only 2 branches, but maybe it would have a benefit with 3 branches, or there's a better way to share the code.

File: tests/baselines/reference/arrayFrom.types [hunk 2]
```diff
@@ -45,18 +45,18 @@ const result1: A[] = Array.from(inputA);
...
 >A : A
 >Array.from(inputA) : A[]
->Array.from : { <T>(iterable: Iterable<T> | ArrayLike<T>): T[]; <T, U>(iterable: Iterable<T> | ArrayLike<T>, mapfn: (v: T, k: number) => U, thisArg?: any): U[]; <T>(arrayLike: ArrayLike<T>): T[]; <T, U>(arrayLike: ArrayLike<T>, mapfn: (v: T, k: number) => U, thisArg?: any): U[]; }
+>Array.from : { <T>(arrayLike: ArrayLike<T>): T[]; <T, U>(arrayLike: ArrayLike<T>, mapfn: (v: T, k: number) => U, thisArg?: any): U[]; <T>(iterable: Iterable<T> | ArrayLike<T>): T[]; <T, U>(iterable: Iterable<T> | ArrayLike<T>, mapfn: (v: T, k: number) => U, thisArg?: any): U[]; }
 >Array : ArrayConstructor
->from : { <T>(iterable: Iterable<T> | ArrayLike<T>): T[]; <T, U>(iterable: Iterable<T> | ArrayLike<T>, mapfn: (v: T, k: number) => U, thisArg?: any): U[]; <T>(arrayLike: ArrayLike<T>): T[]; <T, U>(arrayLike: ArrayLike<T>, mapfn: (v: T, k: number) => U, thisArg?: any): U[]; }
+>from : { <T>(arrayLike: ArrayLike<T>): T[]; <T, U>(arrayLike: ArrayLike<T>, mapfn: (v: T, k: number) => U, thisArg?: any): U[]; <T>(iterable: Iterable<T> | ArrayLike<T>): T[]; <T, U>(iterable: Iterable<T> | ArrayLike<T>, mapfn: (v: T, k: number) => U, thisArg?: any): U[]; }
```
Comment (andy-ms) [inline tests/baselines/reference/arrayFrom.types]: Shouldn't `arrayFrom.ts` have a `Set` test? Then if would break due to this change since `Set` is `Iterable` but not `ArrayLike`.

Comment (rbuckton) [reply tests/baselines/reference/arrayFrom.types]: I'll add a test, but this shouldn't break because there is still an overload that supports `Iterable`. The order just changed because I cleaned up the default order of libs. Previously if you had a `"lib": ["es5", "es2015"]` in your _tsconfig.json_ we would actually end up inserting them in reverse order because we were doing `files.unshift` for default libs.

**Comment by NN---** [comment]
Will this allow fixing #16077 ?

File: src/compiler/commandLineParser.ts [hunk 3]
```diff
@@ -1,6 +1,55 @@
...
+        ["es2018.promise", "lib.es2018.promise.d.ts"],
+        ["es2018.regexp", "lib.es2018.regexp.d.ts"],
+        ["esnext.array", "lib.esnext.array.d.ts"],
+        ["esnext.asynciterable", "lib.esnext.asynciterable.d.ts"],
+    ]);
+
+    // Internally we add some additional lib references that we only support when used as part of a
+    // "lib" reference directive. They are not available on the command line or in tsconfig.json.
+    /* @internal */
+    export const libMap = cloneMap(commandLineLibMap)
```
Comment (mhegazy) [inline src/compiler/commandLineParser.ts]: i would just add this to the map. chances are no one will ever include it.. and if they did.. it does not break anything.

File: src/lib/es2015.full.d.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+/// <reference lib="es5" />
```
Comment (mhegazy) [inline src/lib/es2015.full.d.ts]: should not this just be:

```ts
/// <reference lib="es2015" />
/// <reference lib="dom" />
/// <reference lib="webworker.importscripts" />
/// <reference lib="scripthost" />
/// <reference lib="dom.iterable" />
```

File: src/lib/esnext.full.d.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+/// <reference lib="esnext" />
+/// <reference lib="dom" />
+/// <reference lib="webworker.importscripts" />
+/// <reference lib="scripthost" />
+/// <reference lib="dom.iterable" />
```
Comment (mhegazy) [inline src/lib/esnext.full.d.ts]: nit. i would move this just after `dom`

File: src/lib/webworker.generated.d.ts [hunk 6]
```diff
@@ -1,3 +1,5 @@
+/// <reference lib="webworker.importscripts" />
```
Comment (mhegazy) [inline src/lib/webworker.generated.d.ts]: this is not needed here, the file is autogenerated, and it already has the correct definition. we add it to lib.d.ts to partially allow the lib to be used in webworkers for simple cases.. 

the two files `dom` and `webworker` have too many conflicts in between them, and can not be loaded together in the same context anyways..

so I would just revert the changes to this file, and leave it as is.

File: src/lib/libs.json [hunk 7]
```diff
@@ -46,57 +47,5 @@
         "webworker.generated": "lib.webworker.d.ts",
         "es5.full": "lib.d.ts",
         "es2015.full": "lib.es6.d.ts"
-    },
-    "sources": {
```
Comment (mhegazy) [inline src/lib/libs.json]: can we remove the support for `sources` from jake and gulp as well.

File: src/compiler/commandLineParser.ts [hunk 8]
```diff
@@ -1,6 +1,55 @@
 namespace ts {
     /* @internal */
     export const compileOnSaveCommandLineOption: CommandLineOption = { name: "compileOnSave", type: "boolean" };
+
+    // NOTE: The order here is important to default lib ordering
```
Comment (mhegazy) [inline src/compiler/commandLineParser.ts]: i would expand on this comment and say that the order of the list here is the same order the lib files have in the generated program, and add a note to see use in `createProgram`

**Comment by mhegazy** [comment]
Goto def should work on the new lib references like it does on `/// <reference path="..." />`. take a look at https://github.com/Microsoft/TypeScript/blob/master/src/services/goToDefinition.ts#L136, we will need another check for lib references..

For find-all-refs, i think we can say it does not work on this feature..

File: src/compiler/core.ts [hunk 9]
```diff
@@ -1749,7 +1749,7 @@ namespace ts {
      * Case-sensitive comparisons compare both strings one code-point at a time using the integer
      * value of each code-point after applying `toUpperCase` to each string. We always map both
      * strings to their upper-case form as some unicode characters do not properly round-trip to
-     * lowercase (such as `ẞ` (German sharp capital s)).
+     * lowercase (such as `áºž` (German sharp capital s)).
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply src/compiler/core.ts]: Not sure what caused that to change, but I'll revert.

**Comment by rbuckton** [comment]
I can add "Go To Definition" support, but that requires a number of changes to our Fourslash test harness, as it is currently designed to only use a single lib file.

File: src/services/goToDefinition.ts [hunk 10]
```diff
@@ -89,6 +89,7 @@ namespace ts.GoToDefinition {
     }
 
     export function getReferenceAtPosition(sourceFile: SourceFile, position: number, program: Program): { fileName: string, file: SourceFile } | undefined {
+        debugger;
```
Comment (mhegazy) [inline src/services/goToDefinition.ts]: remove

**Comment by styfle** [comment]
Great job with this PR!

The diff won't load for me so I'll just ask here 🤔 ...

Does this PR actually implement `<reference lib>` in such a way to make the TS engine smaller? Or is this PR just implementing the feature itself and the usage will be implemented in a different PR?

For more context see: https://github.com/Microsoft/TypeScript/issues/23339#issuecomment-380632662

Thanks! 👍

**Comment by mhegazy** [comment]
lib folder should go down by 4 MB or so.

**Comment by RyanCavanaugh** [comment]
@mhegazy @rbuckton  This is going to be a big break to API consumers loading just `lib.d.ts` from the TS and expecting things to work - what's the mitigation here for APIs loading e.g. a fixed set of files from the network on load?

**Comment by weswigham** [comment]
You could always just use an older copy of the lib, if you're one of those API consumers. We ship the latest one with the latest features for consumers, but if you have constraints that force you to, usually you can just include an older copy of the lib in your project - a bunch of our older RWC tests do that.

**Comment by Bnaya** [comment]
Will it add the library to the entire program or only from the reference point and downward?

**Comment by mhegazy** [comment]
not sure i understand the question. the only difference is instead of on file `lib.d.ts` you will have a set of them in your program.

**Comment by mhegazy** [comment]
Looks like we need to ignore `noResolve` when we are resolving lib files as well.. we already have a `ignoreNoDefaultLib` flag, so we can extend this to mean `resolvingLibReference` and ignore noResolve as well as noDefaultLib

**Comment by RyanCavanaugh** [comment]
`--disableIgnoreSkipNoResolve`

**Comment by rbuckton** [comment]
I'll take a look. Lib refs are supposed to ignore noResolve.

**Comment by rbuckton** [comment]
This had nothing to do with `noResolve` and more to do with buildProtocol.ts looking only ignoring `"lib.d.ts"`. It also looks like its already fixed in master.
